src/arb_types.jl

```diff
@@ -1,16 +1,16 @@
-struct mag_struct
+mutable struct mag_struct
```

Comment:
why do we need them `mutable`?

Reply:
That's the reason the `pointer_from_objref` works at all. The object needs to be heap allocated. 

```
  pointer_from_objref(x)

  Get the memory address of a Julia object as a Ptr. The existence of the resulting Ptr will not protect the object
  from garbage collection, so you must ensure that the object remains referenced for the whole time that the Ptr will
  be used.

  This function may not be called on immutable objects, since they do not have stable memory addresses.

  See also: unsafe_pointer_to_objref.
```

But this changes the memory layout of our structs since it now contains another pointer and thus a second indirection :/

Reply:
I think it needs to be to be able to use it in a `ccall`, but I'm not entirely sure. Removing it breaks things at least, so it is needed in some way.

Reply:
I think it's fine to have these mutable, bu then we should also attach the finalizer here and change `Arb` etc. to be a `struct`.

Reply:
but why do we need to get a pointer by calling `pointer_from_objref`? `ccall` doesn't use it, it goes through `cconvert` and `unsafe_convert`

Reply:
We don't need to use `pointer_from_objref`, so far I have only used it in some tests. But the reason that `pointer_from_objref` doesn't work on immutable data is likely the same reason that `ccall` cannot handle immutable objects.

Reply:
well, I probably don't understand what do you mean by saying that `ccall` can handle immutable data.
(the following is happening on `master`, i.e. immutable `arb_struct`):
```julia
julia> a = Arb("0.1")
[0.1000000000000000000000000000000000000000000000000000000000000000000000000000 +/- 1.95e-78]

julia> b = zero(a)
0

julia> rb = Base.cconvert(Ref{Arblib.arb_struct}, b.arb)
Base.RefValue{Arblib.arb_struct}(Arblib.arb_struct(0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000))

julia> ccall(Arblib.@libarb(arb_add), Cvoid, (Ref{Arblib.arb_struct}, Ref{Arblib.arb_struct}, Ref{Arblib.arb_struct}), rb, a.arb, a.arb)

julia> rb[] #rb was modified since it was passed by reference 
Arblib.arb_struct(0xfffffffffffffffe, 0x0000000000000008, 0x0000000000000004, 0x0000000000eaee40, 0xfffffffffffffeff, 0x0000000020000000)

julia> b.arb
Arblib.arb_struct(0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000)

julia> Arblib.add!(b,a,a)

julia> ccall(Arblib.@libarb(arb_equal), Cint, (Ref{Arblib.arb_struct}, Ref{Arblib.arb_struct}), rb[], zero(a).arb)
0

julia> ccall(Arblib.@libarb(arb_equal), Cint, (Ref{Arblib.arb_struct}, Ref{Arblib.arb_struct}), rb[], b.arb)
1 # rb[] and b.arb are equal

```

Reply:
But in this case the `arb_struct` is inlined in the `Arb` which is mutable. But I don't really know how this works.

Reply:
I think all this only works, since `b` is heap allocated, so you have a stable memory address for `b.arb`. If however, you would construct an `arb_struct` from scratch, then it would be only be living in the current function stack. And then you cannot pass this to `ccall` without risking a segfault in my understanding.